### PR TITLE
feat: [DENA-582] adding release-please

### PR DESCRIPTION
**Pull request by MarcinGinszt**
We discussed the possible release naming for shared-kustomize-bases.

We considered using commit hashes (like [system-manifests](https://github.com/utilitywarehouse/system-manifests)), but we decided to go with semver for it's benefits and clearence. 

While using release-please, we can either use simple semver, or consider the image version in it.

In order to consider image version in it, we would have to keep separate image versions in separate directories- like `redis-1.2.3`. 
In order to minimize amount of versions, we could just keep the separate directories for major image releases- like `redis-1`.

With this approach, we could maintain new versions of manifests for older image versions. However, the downside would be maintenance overhead.

Since we didn't encountered a problem with old image version demanding separate manifests releases, we will just go with the simple semver. In case we'll need it in the future, we can just add new directories for desired image versions.

### CODE REVIEW

File: .release-please-manifest.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "cockroachdb": "1.0.0",
+  "elasticsearch": "1.0.0",
+  "es-topic-indexer": "1.0.0",
+  "redis": "1.0.0",
+  "redis-cluster": "1.0.0"
```
Comment (MarcinGinszt) [inline .release-please-manifest.json:6]: Since we are forced to use semver with release-please, I think we can start from 1.0.0.
Before that, we used versions like `"image version tag"-1` for cockroach manifest.

Comment (matthewhughes-uw) [reply .release-please-manifest.json:6]: > we used versions like "image version tag"-1 for cockroach manifest.

Ah right, now I remember, trying to match this format was the painful bit when trying to use `release-please` in the cockroachdb manifests. Do we still need to surface the version of the underlying tech somewhere? E.g. if someone is running `vX` of cockroachDB how would they know which version of the manifests in this repo are compatible with that? Maybe that information just lives in the changelog?

Comment (MarcinGinszt) [reply .release-please-manifest.json:6]: >  Maybe that information just lives in the changelog?

Yeah, I agree with that. Doesn't seem like anyone would like to maintain new versions of manifests for the old version of the image - and even if this situation will happen, I don't think we should base our release system on that. 

I'm kind of uneasy with versioning all the existing software as 1.0.0, though. But I don't know which version else could we use.

Comment (sbuliarca) [reply .release-please-manifest.json:6]: Indeed, I just found the slack convo https://utilitywarehouse.slack.com/archives/C03LB7ZN07J/p1718869775335839?thread_ts=1718781427.663029&cid=C03LB7ZN07J

TLDR we were thinking that something like `redis-v7.2.0-base-v1.0.0` would work.

Comment (matthewhughes-uw) [reply .release-please-manifest.json:6]: > TLDR we were thinking that something like redis-v7.2.0-base-v1.0.0 would work.

I don't think that format works well with `release-please` https://github.com/utilitywarehouse/kafka-manifests/pull/41. If we don't plan on supporting more than one version of each tech (IIRC from recent discussions, e.g. we don't want manifests for both Redis v7 and Redis v6) do we need/want this tag format (basically: do we want to go through the extra effort to try and make that tag format work with `release-please`?)

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  release-please-release:
+    # match on the default PR title
+    # https://github.com/googleapis/release-please/blob/3569f158158e2bd096a94c145bdb1a7f455829f0/docs/customizing.md#pull-request-title
+    if: "${{ github.event_name == 'push' && contains(github.event.head_commit.message, 'chore(main): release') }}"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: google-github-actions/release-please-action@v4
+        with:
+          token: ${{secrets.GITHUB_TOKEN}}
+          skip-github-pull-request: true
```
Comment (matthewhughes-uw) [inline .github/workflows/release.yml:28]: I think we can keep this file simple for now. I don't expect the same volume of changes as in the uwos-go repo (which is the need for the complexity there: https://github.com/utilitywarehouse/uwos-go/commit/e6b168bd1762125893ca85be577672a2f137bdbe)

```suggestion
name: Release
on:
  push:
    branches:
      - main
jobs:
  release-please:
    runs-on: ubuntu-latest
    steps:
      - uses: google-github-actions/release-please-action@v4
        with:
          token: ${{secrets.GITHUB_TOKEN}}
```

Comment (MarcinGinszt) [reply .github/workflows/release.yml]: [Applied](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/41/commits/6ff6fbc4f01c1f74b51d39e1f9b57aee4a8d24eb)

**Comment by matthewhughes-uw** [comment]
It looks like a bunch of extra changes have snuck in (which I assume is what's causing the failing action)

File: .release-please-manifest.json (hunk 1 above)
Comment (sbuliarca) [reply .release-please-manifest.json:6]: Here's a summary of our sync discussion:
- we're going with offering a "full" product package for now with keeping the tags as component-x.y.z without exposing the component version in the tag. 
- The release changelog will specify which version of the component is supported in that release
- we're gonna present this option at TAG and see what people think, and we can adjust afterwards

**Comment by MarcinGinszt** [comment]
@matthewhughes-uw 
Pipeline is failing, because there is this file cockroachbdb/kustomization.yaml, referencing non-existing files on main branch. I guess we should have a second look on our pipeline- it was not reported on https://github.com/utilitywarehouse/shared-kustomize-bases/pull/39

File: .release-please-manifest.json (hunk 1 above)
Comment (sbuliarca) [reply .release-please-manifest.json:6]: :) didn't notice that

File: .github/workflows/pr.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  main:
+    name: Validate PR title
+    runs-on: ubuntu-latest
+    steps:
+      - uses: amannn/action-semantic-pull-request@v5
+        id: lint_pr_title
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          # https://github.com/utilitywarehouse/uwos-go/blob/main/release-please-config.json#L53
```
Comment (matthewhughes-uw) [inline .github/workflows/pr.yml:18]: comment points at the wrong repo

File: release-please-config.json [hunk 4]
```diff
@@ -0,0 +1,79 @@
+{
+  "bootstrap-sha": "77ecc2590ea9899db4f47a92ca1bbb50c18c6777",
```
Comment (matthewhughes-uw) [inline release-please-config.json:2]: I don't think we want this, it doesn't point to a commit in this repo. We could probably try the commit on `main` before this PR is merged, docs: https://github.com/googleapis/release-please/blob/main/docs/manifest-releaser.md#bootstrapping or just try without this value and see what happens :shrug:

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+name: Release
+on:
+  push:
+    branches:
+      - main
+jobs:
+  release-please:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: google-github-actions/release-please-action@v4
```
Comment (matthewhughes-uw) [inline .github/workflows/release.yml:10]: This action is deprecated since last month, we should try out the new one

```suggestion
      - uses: googleapis/release-please-action@v4
```

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+        with:
+          token: ${{secrets.GITHUB_TOKEN}}
```
Comment (matthewhughes-uw) [inline .github/workflows/release.yml:12]: :thinking: does the token by default have permissions to raise PRs etc.? The basic example (https://github.com/googleapis/release-please-action?tab=readme-ov-file#basic-configuration) uses:

```yaml

permissions:
  contents: write
  pull-requests: write
```

(but then goes on to use `${{ secrets.MY_RELEASE_PLEASE_TOKEN }}` which makes those permissions unused anyway :upside_down_face:)

But in `uwos-go` we never set these, so I guess not?

**Review by matthewhughes-uw — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yml [hunk 7]
```diff
@@ -0,0 +1,53 @@
+name: PR
+on:
+  pull_request_target:
```
[the remark this answers is not in the document]
Comment (MarcinGinszt) [reply .github/workflows/pr.yml:3]: Good question, as always :) 
I took this workflow from the [uwos-go](https://github.com/utilitywarehouse/uwos-go/blob/main/.github/workflows/pr.yml).
It makes the workflow run against the target branch of the repository, while granting all the secrets. This sort of trigger was actually added to allow secure run of workflows with secrets in the public repos. 
I suppose here it was used to make the workflow [run correctly for forks](https://stackoverflow.com/questions/74957218/what-is-the-difference-between-pull-request-and-pull-request-target-event-in-git). I don't expect many external contributions, but cool detail, I think.
The article you linked points out it might be dangerous, but
```
Generally speaking, when the PR contents are treated as passive data, i.e. not in a position of influence over the build/testing process, it is safe. But the repository owners must be extra careful not to trigger any script that may operate on PR controlled contents like in the case of `npm install`.
```

I don't mind changing it to `pull_requets`, even though it should be safe now. 
I can also just add a comment with warning to this file, to prevent dangerous contributions. 

What would you suggest?

[the remark this answers is not in the document]
Comment (matthewhughes-uw) [reply .github/workflows/pr.yml:3]: > I don't mind changing it to pull_requets, even though it should be safe now.

I like the idea of `pull_requests` mostly because it's used most other places (that I've seen/can remember) and also as a best-practice

File: .github/workflows/release.yml (hunk 5 above)
Comment (MarcinGinszt) [reply .github/workflows/release.yml]: [Applied](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/41/commits/0262f3fa8ab3c0367a8a4ca97d1bc719ec894307#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R10)

File: .github/workflows/pr.yml (hunk 3 above)
Comment (MarcinGinszt) [reply .github/workflows/pr.yml]: [I removed the comment](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/41/commits/0262f3fa8ab3c0367a8a4ca97d1bc719ec894307#diff-15c806aa509538190832852f439e9921a23bec2da81f95ed0e4bf13c14e5b160L18) since I don't know what should it signify

File: .github/workflows/pr.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (MarcinGinszt) [reply .github/workflows/pr.yml]: [Applied](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/41/commits/0262f3fa8ab3c0367a8a4ca97d1bc719ec894307#diff-15c806aa509538190832852f439e9921a23bec2da81f95ed0e4bf13c14e5b160R3)

File: .github/workflows/pr.yml (hunk 3 above)
Comment (matthewhughes-uw) [reply .github/workflows/pr.yml]: > [I removed the comment](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/41/commits/0262f3fa8ab3c0367a8a4ca97d1bc719ec894307#diff-15c806aa509538190832852f439e9921a23bec2da81f95ed0e4bf13c14e5b160L18) since I don't know what should it signify

I think it's noting that this list is based off the entries in the `changelog-sections` in the release please config. If we wanted to we could probably just add a step to generate this list from the config file, but I imagine the list is not going to change much so :shrug: probably not worth the effort

File: .github/workflows/release.yml (hunk 6 above)
Comment (MarcinGinszt) [reply .github/workflows/release.yml:12]: Yeah, I think if this org-wide secret is good enough for uwos-go, it should be good enough for shared-kustomize-bases :slightly_smiling_face:

Comment (matthewhughes-uw) [reply .github/workflows/release.yml:12]: > Yeah, I think if this org-wide secret is good enough for uwos-go, it should be good enough for shared-kustomize-bases 🙂

It's not an org-wide secret, it's the one GitHub gives us in each action run, but looking at some docs https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token I assume our org has things setup in "Default access(permissive)" mode so we get the permissions we need by default

**Review by matthewhughes-uw — APPROVED** [review]
